### ISSUE TITLE
Unencrypted Databases getting Encrypted by Mistake?

### ISSUE THREAD

**Issue by DavidKann** [sqlitebrowser/sqlitebrowser#1314]
### Details for the issue

I have several students using SQLite Browser and a few of them have come to me because their databases have become encrypted. They don't have any recollection of having set up encryption and I've sat with them while they saved new databases only to find them encrypted when they try to open them again. Decrypting with a blank password is ineffective. I've inspected their database files with a texteditor and the contents do appear to be gibberish, although I don't know if that's because it's encrypted, or because the data isn't plaintext:

![image](https://user-images.githubusercontent.com/34763715/35840151-5659124c-0b2f-11e8-852a-b842636472dc.png)

I was hoping to get some advice about how they could either recover their databases or prevent this issue from reoccurring. Thank you very much for your advice!

### Useful extra information

We are running OSX variations including Sierra and High Sierra and SQLite Browser 3.10.1
Only some students have had this problem. I haven't seen it on my own system, nor have most other students.

#### I'm opening this issue because:

- [ ] DB4S is crashing
- [x] DB4S has a bug
- [ ] DB4S needs a feature
- [ ] DB4S has another problem

#### I'm using DB4S on:

- [ ] Windows: ( _version:_ ___ )
- [ ] Linux: ( _distro:_ ___ )
- [X] Mac OS: ( 10 version:13.2)
- [ ] Other: ___

#### I'm using DB4S version:

- [x] 3.10.1
- [ ] 3.10.0
- [ ] 3.9.1
- [ ] Other: ___

#### I have also:

- [ ] Tried out the latest nightly version: https://github.com/sqlitebrowser/sqlitebrowser#nightly-builds
- [x] Searched for an existing similar issue: https://github.com/sqlitebrowser/sqlitebrowser/issues?utf8=%E2%9C%93&q=is%3Aissue%20

**Comment by chrisjlocke** [comment]
Are there any backups of the file?  Is it always the same users?  Are the files stored on a server, or locally?  
DB4S doesn't normally encrypt or decrypt automatically, so the corruption is unusual.  DB4S is used fairly regularly, and we haven't had this reported before, so its even more unusual.

I note the 'official release' for OSX is just one binary, while the nightlies are split with either having SQLCipher or not.  If you don't require encryption support, try one of the nightlies?  

The database you've examined doesn't even look like a corrupted database - there isn't even fragments of a database.  DB4S is usually pretty sensible, and won't open a 'normal' database by trying to open an 'encrypted' database.  You can't even open a normal database and save it as an encrypted database 'accidentally'.  

If some file I/O was corrupting the database (which has been known on network drives, which is why I asked if it was local or on a network) it wouldn't have corrupted/encrypted as much as your example.  From that fragment, it does appear encrypted rather than corrupted.

Its a bizarre one.

**Comment by justinclift** [comment]
Interesting.  This does indeed sound weird.

@DavidKann The "encrypted" thing might be a false lead, as @chrisjlocke is mentioning.

When DB4S (our abbreviation for DB Browser for SQLite) is given a database file to load, it first tries loading it using the SQLite library.

If that library returns an error to DB4S along the lines of "not a database file" then we assume we've been passed an encrypted file and proceed to ask for the password.  Then try opening it again, but using the SQLCipher library (encryption) this time.

That second sequence, where DB4S prompts for the encryption password, can throw people off.  If the supplied file isn't valid database data for some reason, the encryption prompt that pops up can be confusing. :wink:

---

As well as the query about network storage possibly corrupting things (rare, but it can happen), another avenue to look at is whether the wrong SQLite library is being loaded by OSX somehow.

What is the output of `otool -L` on one of the systems that's saving incorrectly?

On a OSX (El Capitan) system here, with a freshly downloaded nightly build (from [this one exactly](https://nightlies.sqlitebrowser.org/osx/DB%20Browser%20for%20SQLite_20180206.dmg)), the output is:

```
$ otool -L /Applications/DB\ Browser\ for\ SQLite.app/Contents/MacOS/DB\ Browser\ for\ SQLite 
/Applications/DB Browser for SQLite.app/Contents/MacOS/DB Browser for SQLite:
	@executable_path/../Frameworks/libsqlite3.0.dylib (compatibility version 9.0.0, current version 9.6.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1238.0.0)
	/System/Library/Frameworks/Carbon.framework/Versions/A/Carbon (compatibility version 2.0.0, current version 157.0.0)
	@rpath/QtPrintSupport.framework/Versions/5/QtPrintSupport (compatibility version 5.9.0, current version 5.9.2)
	@rpath/QtWidgets.framework/Versions/5/QtWidgets (compatibility version 5.9.0, current version 5.9.2)
	@rpath/QtGui.framework/Versions/5/QtGui (compatibility version 5.9.0, current version 5.9.2)
	@rpath/QtCore.framework/Versions/5/QtCore (compatibility version 5.9.0, current version 5.9.2)
	/System/Library/Frameworks/DiskArbitration.framework/Versions/A/DiskArbitration (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)
	@rpath/QtOpenGL.framework/Versions/5/QtOpenGL (compatibility version 5.9.0, current version 5.9.2)
	@rpath/QtMacExtras.framework/Versions/5/QtMacExtras (compatibility version 5.9.0, current version 5.9.2)
	@rpath/QtNetwork.framework/Versions/5/QtNetwork (compatibility version 5.9.0, current version 5.9.2)
	@rpath/QtConcurrent.framework/Versions/5/QtConcurrent (compatibility version 5.9.0, current version 5.9.2)
	@rpath/QtXml.framework/Versions/5/QtXml (compatibility version 5.9.0, current version 5.9.2)
	/System/Library/Frameworks/OpenGL.framework/Versions/A/OpenGL (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/AGL.framework/Versions/A/AGL (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 307.4.0)
```

That output shows `libsqlite3.0.dylib` in a `Frameworks` subdirectory relative to the executable file (inside the .app package).   If yours is different, that could be where the problem is coming from.

**Comment by DavidKann** [comment]
Hi Chris and Justin,

The effected database files are local files, no networking involved. The students were not keeping backups of their files, unfortunately. 

For the mean time I'll direct students to the nightly build without encryption

Next time we have class, I'll get the students to try running otool. For what it's worth, here is the output on my machine, but I haven't run into the error myself. I notice that I don't have libsqlite3.0.dylib as you do, Justin, but libsqlcipher.0.dylib

> otool -L /Applications/DB\ Browser\ for\ SQLite.app/Contents/MacOS/DB\ Browser\ for\ SQLite
/Applications/DB Browser for SQLite.app/Contents/MacOS/DB Browser for SQLite:
	@executable_path/../Frameworks/libsqlcipher.0.dylib (compatibility version 9.0.0, current version 9.6.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1238.0.0)
	/System/Library/Frameworks/Carbon.framework/Versions/A/Carbon (compatibility version 2.0.0, current version 157.0.0)
	@rpath/QtPrintSupport.framework/Versions/5/QtPrintSupport (compatibility version 5.7.0, current version 5.7.1)
	@rpath/QtWidgets.framework/Versions/5/QtWidgets (compatibility version 5.7.0, current version 5.7.1)
	@rpath/QtGui.framework/Versions/5/QtGui (compatibility version 5.7.0, current version 5.7.1)
	@rpath/QtCore.framework/Versions/5/QtCore (compatibility version 5.7.0, current version 5.7.1)
	/System/Library/Frameworks/DiskArbitration.framework/Versions/A/DiskArbitration (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)
	@rpath/QtMacExtras.framework/Versions/5/QtMacExtras (compatibility version 5.7.0, current version 5.7.1)
	@rpath/QtNetwork.framework/Versions/5/QtNetwork (compatibility version 5.7.0, current version 5.7.1)
	/System/Library/Frameworks/OpenGL.framework/Versions/A/OpenGL (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/AGL.framework/Versions/A/AGL (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 307.4.0)

**Comment by justinclift** [comment]
> I notice that I don't have libsqlite3.0.dylib as you do, Justin, but libsqlcipher.0.dylib

Ahhh yean.  That just means you're running a version that was built with SQLCipher.  Nothing looks wrong there. :smile:

**Comment by chrisjlocke** [comment]
>  I notice that I don't have libsqlite3.0.dylib as you do

I suspect that might be because the 'official release' is built using SQLCipher, while the nightlies are optional.  Either way, you're using the SQLite thats with DB4S, so thats a good sign.

Thanks for raising this issue though, rather than abandoning the software!  Interesting you're using it in a class environment.  

> and I've sat with them while they saved new databases only to find them encrypted when they try to open them again.

Is this reproducible?  A new database is created, table added, fields added, database closed, and the database is 'encrypted' when next opened?
It would be handy to see what would happen if a 'known to be OK' database was opened and saved on one of the troublesome machines.

**Comment by DavidKann** [comment]
> Thanks for raising this issue though, rather than abandoning the software! Interesting you're using it in a class environment.

No worries! DB4S has a straightforward interface. Once the students get a primer on what a database is, they don't have much trouble learning the environment. It's a great tool and I'm very happy to be able to use it.

> Is this reproducible? A new database is created, table added, fields added, database closed, and the database is 'encrypted' when next opened?
It would be handy to see what would happen if a 'known to be OK' database was opened and saved on one of the troublesome machines.

I've never reproduced it on my machine, but next time I see one of the students with a currupted database I'll get them to run otool and we'll try to reproduce the error. We'll also try to pass them a database that I made on my machine as well as try the nightly without encryption. You'll have to forgive me while I make you wait for a few days though, because I don't see my students every day.

Thank you again for your advice!

**Comment by chrisjlocke** [comment]
> You'll have to forgive me while I make you wait for a few days though

No problem.  We haven't had reports of this from anyone else, so its some weird situation occurring for you.  DB4S is careful on what it opens - it ensures its a valid database - and SQLite itself is very robust and journalistic, so whats happening shouldn't be able to happen!

**Comment by DavidKann** [comment]
I met with one of the students who was affected. We ran otool and this is the output:

	@executable_path/../Frameworks/libsqlcipher.0.dylib (compatibility version 9.0.0, current version 9.6.0)

	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1238.0.0)

	/System/Library/Frameworks/Carbon.framework/Versions/A/Carbon (compatibility version 2.0.0, current version 157.0.0)

	@rpath/QtPrintSupport.framework/Versions/5/QtPrintSupport (compatibility version 5.7.0, current version 5.7.1)

	@rpath/QtWidgets.framework/Versions/5/QtWidgets (compatibility version 5.7.0, current version 5.7.1)

	@rpath/QtGui.framework/Versions/5/QtGui (compatibility version 5.7.0, current version 5.7.1)

	@rpath/QtCore.framework/Versions/5/QtCore (compatibility version 5.7.0, current version 5.7.1)

	/System/Library/Frameworks/DiskArbitration.framework/Versions/A/DiskArbitration (compatibility version 1.0.0, current version 1.0.0)

	/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)

	@rpath/QtMacExtras.framework/Versions/5/QtMacExtras (compatibility version 5.7.0, current version 5.7.1)

	@rpath/QtNetwork.framework/Versions/5/QtNetwork (compatibility version 5.7.0, current version 5.7.1)

	/System/Library/Frameworks/OpenGL.framework/Versions/A/OpenGL (compatibility version 1.0.0, current version 1.0.0)

	/System/Library/Frameworks/AGL.framework/Versions/A/AGL (compatibility version 1.0.0, current version 1.0.0)

	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 307.4.0)

Meanwhile, we installed the nightly without encryption on his machine. I imagine that whatever the problem was (program error or user error) if DB4S can't encrypt the database, it won't become encrypted. With the next round of students I'll recommend them to use the nightly without encryption

**Comment by chrisjlocke** [comment]
Just bear in mind that the nightlies contain 'experimental' and 'unfinished' code, so _possibly_ is more unreliable than the proper released versions.  It would obviously be handy to identify the cause of the issue ... while you were kind enough to report the issue, other people might have thrown the software away in disgust and never told us, so the underlying cause goes unresolved... However, I appreciate that its not easy to spend time doing this.
The initial output from otool doesn't seem to shed much light.  :(

**Comment by justinclift** [comment]
Yep, that otool output isn't showing anything weird.  Would have been nice if there was a smoking gun there (eg wrong library being loaded), but oh well.

It'll be interesting to see if the problem still happens for these students using the nightly builds.  I'm half expecting it will still happen... kind of (personally) suspecting that whatever was going on was just mucking up the saving somehow and not really encrypting things.  Guess we'll see what happens with the next few classes. :smile:

**Comment by DavidKann** [comment]
Sorry, we weren't able to reproduce the error again today. I'll talk tomorrow with the other student who was having the problem. However the rest of the class has moved on so I won't have any more data until next year, unfortunately. I'll update you tomorrow with the results.

**Comment by justinclift** [comment]
No worries. :smile:

**Comment by chrisjlocke** [comment]
> kind of (personally) suspecting that whatever was going on was just mucking up the saving somehow

Unless its OS specific, I'm surprised this issue hasn't cropped up earlier, if it was program related... I use DB4S almost daily (albeit it mostly doing similar tasks) so 'should' have had a corrupted file by now.

**Comment by justinclift** [comment]
Yeah.  My suspicion was that something had been installed on the computers which changed the library loading order, causing an incorrect library to be loaded.  The `otool -L` output isn't showing that behaviour though. 😕

**Comment by chrisjlocke** [comment]
In a sort of related matter, someone has posted to the SQLite mailing list about the headers of a database becoming corrupted.  This has occurred more of programming directly with SQLite (ie, using it in your own applications) than how DB4S uses it (ie, just sending it database manipulation commands).
However, it has raised an interesting query.  If the file header was replaced, would the rest of the file be OK, or is it indeed encrypted.

@DavidKann - would it be possible to send me a corrupted/encrypted database which you can no longer open?
You can email it to encrypted [at] chrisjlocke [dot] co [dot] uk.

No problem if you don't have access to the database anymore ... was just a wild hunch.

**Comment by DavidKann** [comment]
Hi Chris,

That's an interesting idea.It would be helpful to know if the database was indeed encrypted. If it is, then switching to the nightly without SQCipher should do the trick.

I don't think there are any issues with posting the suspect database in this thread so if it's all the same do you, I'll just attach it here (zipped to get it attached to the message).

Cheers!
David
[internal assessment.db.zip](https://github.com/sqlitebrowser/sqlitebrowser/files/1709342/internal.assessment.db.zip)

**Comment by chrisjlocke** [comment]
Well, that didn't work.  I didn't think it would do, as (normally) a SQLite database still has human-readable data in it, while your data is totally garbled.
:-(

**Comment by DavidKann** [comment]
That's too bad. I'm sorry I wasn't able to get you the information you needed to solve this problem. 

Next year I'll try recommending the nightly build without encryption features to my students and we'll see what comes of it. In the mean time I just wanted to say thank you to you and Justin for an awesome piece of software!

**Comment by justinclift** [comment]
You're welcome David.  It's most definitely a group effort. 😁

**Comment by justinclift** [comment]
Ahhh.  That sounds like something went wrong when the file was closed, so the "journal" file wasn't deleted.

One thing to try (after taking a backup of both files), is to just delete the `project.db-journal` file when DB4S isn't running.  Then start DB4S up and open the `project.db` file.  **If** the file is actually in a decent state that **should** (in theory) "just work".  If it doesn't though... ugh. :frowning:
